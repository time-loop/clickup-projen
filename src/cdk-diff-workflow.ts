import { SampleFile, YamlFile, javascript } from 'projen';
import { NodePackage } from 'projen/lib/javascript';
import { clickupCdk } from './clickup-cdk';
import { OptionalNodeVersion } from './optional-node-version';
import { parameters } from './utils/parameters';

export module cdkDiffWorkflow {
  function createCdkDiffWorkflow(options: CDKDiffOptionsConfig) {
    const installDeps =
      options.packageManager === javascript.NodePackageManager.PNPM
        ? 'pnpm i --frozen-lockfile'
        : 'yarn install --check-files';

    const installPnpm = {
      name: 'Setup pnpm',
      uses: 'pnpm/action-setup@v3',
      with: {
        version: options.pnpmVersion ?? '9',
      },
    };

    const defaultWorkflow = {
      name: 'cdk-diff',
      on: {
        pull_request: {},
        workflow_dispatch: {},
      },
      permissions: {
        'id-token': 'write',
        contents: 'write',
        'pull-requests': 'write',
      },
      jobs: {
        diff: {
          'runs-on': 'ubuntu-latest',
          env: {
            CI: 'true',
            REPO_NAME: '${{ github.event.repository.name }}',
          },
          steps: [
            {
              name: 'Checkout',
              uses: 'actions/checkout@v3',
              with: {
                ref: '${{ github.event.pull_request.head.ref }}',
                repository: '${{ github.event.pull_request.head.repo.full_name }}',
              },
            },
            {
              name: 'GitHub Packages authorization',
              run: [
                'cat > .npmrc <<EOF',
                '//npm.pkg.github.com/:_authToken=${{ secrets.ALL_PACKAGE_READ_TOKEN }}',
                '@time-loop:registry=https://npm.pkg.github.com/',
                'EOF',
              ].join('\n'),
            },
            {
              name: 'Setup Node.js',
              uses: 'actions/setup-node@v3',
              with: {
                'node-version': options.nodeVersion ?? parameters.PROJEN_NODE_VERSION,
              },
            },
            ...(options.packageManager === javascript.NodePackageManager.PNPM ? [installPnpm] : []),
            {
              name: 'Install dependencies',
              run: installDeps,
            },
            createStackCaptureStep(options.envsToDiff),
            ...createDiffSteps(options.envsToDiff).flat(),
            createCdkNotifierStep(options.envsToDiff),
            ...createLabelRemovalSteps(options.envsToDiff),
            ...createLabelSteps(options.envsToDiff),
            {
              name: 'Should auto merge',
              id: 'should-auto-merge',
              if: `${getAutoMergeConditions(
                options.envsToDiff,
              )} && github.event.pull_request.user.login == 'cu-infra-svc-git'`,
              run: 'echo "auto-merge=true" >> $GITHUB_OUTPUT',
            },
            {
              name: 'Enable automerge',
              uses: 'daneden/enable-automerge-action@v1.0.2',
              if: 'steps.should-auto-merge.outputs.auto-merge',
              with: {
                'github-token': '${{ secrets.PROJEN_GITHUB_TOKEN }}',
                'merge-method': 'SQUASH',
                'allowed-author': 'cu-infra-svc-git',
              },
            },
            {
              name: 'Auto approve',
              uses: 'hmarr/auto-approve-action@v2.2.1',
              if: 'steps.should-auto-merge.outputs.auto-merge',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
              },
            },
            {
              name: 'Save processed diff logs',
              uses: 'actions/upload-artifact@v4',
              with: {
                name: 'ProcessedDiffLogs',
                path: '*.log',
                'retention-days': 3,
              },
            },
          ],
        },
      },
    };

    return defaultWorkflow;
  }
  function getStacksEnvVarForGivenEnv(envName: string) {
    return `STACKS_TO_DIFF_${envName.toUpperCase()}`;
  }

  function createStackCaptureStep(envsToDiff: (EnvToDiff | ExplicitStacksEnvToDiff)[]) {
    return {
      name: 'capture stacks to diff',
      run: [
        './node_modules/.bin/cdk ls -l -j --notices=false > cdk-ls.json',
        ...createEnvVarsToCaptureStacksToDiff(envsToDiff),
      ].join('\n'),
    };
  }

  function createEnvVarsToCaptureStacksToDiff(envsToDiff: (EnvToDiff | ExplicitStacksEnvToDiff)[]) {
    const stacksToDiff = envsToDiff.map((env) => {
      const envVar = getStacksEnvVarForGivenEnv(env.name);
      if ('stackSearchString' in env) {
        return `${envVar}=$(jq -r \'[.[].id | select(contains("${env.stackSearchString}")) ] | join(" ")\' cdk-ls.json)\necho "${envVar}=$${envVar}" >> $GITHUB_ENV`;
      } else {
        return `${envVar}="${env.stacks.join(' ')}"\necho "${envVar}=$${envVar}" >> $GITHUB_ENV`;
      }
    });

    return stacksToDiff;
  }

  function createDiffSteps(envsToDiff: (EnvToDiff | ExplicitStacksEnvToDiff)[]) {
    return envsToDiff.map((env) => {
      const envVar = getStacksEnvVarForGivenEnv(env.name);
      return [
        {
          name: `configure ${env.name} aws credentials`,
          uses: 'aws-actions/configure-aws-credentials@v1',
          with: {
            'role-to-assume': env.oidcRoleArn,
            'role-duration-seconds': env.roleDuration ?? 900,
            'aws-region': 'us-west-2',
          },
        },

        {
          name: `diff ${env.name}`,
          run: `set -o pipefail\n./node_modules/.bin/cdk diff $${envVar} --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-${
            env.name
          }.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/lib/cdkLogParser.js cdk-${
            env.name
          }.log && echo '${env.name.toUpperCase()}_HAS_NO_DIFF=true' >> $GITHUB_ENV || echo 'Diffs present!'\n`,
        },
      ];
    });
  }

  function createCdkNotifierStep(envsToDiff: (EnvToDiff | ExplicitStacksEnvToDiff)[]) {
    return {
      name: 'cdk-notify',
      run: [...createDownloadCdkNotifierStep(), ...createCdkNotifierSteps(envsToDiff)].join('\n'),
    };
  }

  function createDownloadCdkNotifierStep() {
    return [
      'curl -L "https://github.com/karlderkaefer/cdk-notifier/releases/download/v2.3.2/cdk-notifier_$(uname)_amd64.gz" -o cdk-notifier.gz',
      'gunzip cdk-notifier.gz && chmod +x cdk-notifier && rm -rf cdk-notifier.gz',
    ];
  }

  function createCdkNotifierSteps(envsToDiff: (EnvToDiff | ExplicitStacksEnvToDiff)[]) {
    return envsToDiff.map((env) => {
      return `./cdk-notifier --log-file ./cdk-${env.name}.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token \${{ secrets.GITHUB_TOKEN }} --pull-request-id \${{ github.event.pull_request.number }} -t '${env.name} Stacks - If truncated please see the action for the full log!'`;
    });
  }

  function createLabelSteps(envsToDiff: (EnvToDiff | ExplicitStacksEnvToDiff)[]) {
    return envsToDiff.map((env) => {
      return {
        name: `Apply '${env.labelToApplyWhenNoDiffPresent}' label based on diff status`,
        if: `env.${env.name.toUpperCase()}_HAS_NO_DIFF == 'true'`,
        uses: 'actions/github-script@v6',
        with: {
          'github-token': '${{ secrets.GITHUB_TOKEN }}',
          script: [
            'github.rest.issues.addLabels({',
            'issue_number: context.issue.number,',
            'owner: context.repo.owner,',
            'repo: context.repo.repo,',
            `labels: ["${env.labelToApplyWhenNoDiffPresent}"]`,
            '})',
          ].join('\n'),
        },
      };
    });
  }

  function createLabelRemovalSteps(envsToDiff: (EnvToDiff | ExplicitStacksEnvToDiff)[]) {
    return envsToDiff.map((env) => {
      return {
        name: `Remove '${env.labelToApplyWhenNoDiffPresent}' label based on diff status given a previous commit may have had no diff but the current one does`,
        if: `env.${env.name.toUpperCase()}_HAS_NO_DIFF != 'true'`,
        uses: 'actions/github-script@v6',
        with: {
          'github-token': '${{ secrets.GITHUB_TOKEN }}',
          script: [
            'const labels = await github.paginate(',
            'github.rest.issues.listLabelsOnIssue, {',
            'issue_number: context.issue.number,',
            'owner: context.repo.owner,',
            'repo: context.repo.repo,',
            '}',
            ')',
            `if (labels.find(label => label.name === "${env.labelToApplyWhenNoDiffPresent}")) {`,
            'github.rest.issues.removeLabel({',
            'owner: context.repo.owner,',
            'repo: context.repo.repo,',
            'issue_number: context.issue.number,',
            `name: "${env.labelToApplyWhenNoDiffPresent}"`,
            '})',
            '}',
          ].join('\n'),
        },
      };
    });
  }

  function getAutoMergeConditions(envsToDiff: (EnvToDiff | ExplicitStacksEnvToDiff)[]) {
    return envsToDiff.map((env) => `env.${env.name.toUpperCase()}_HAS_NO_DIFF == 'true'`).join(' && ');
  }

  interface BaseEnvToDiff {
    /**
     * Unique name for the cdk diff action
     *
     * This will be used to create the output file name, header on comments, and more
     * Example: `qa`, `staging`, `prod`
     */
    readonly name: string;

    /**
     * Name of the OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcRoleArn: string;

    /**
     * Label that will be applied to the PR when there are no changes in the diff
     * Example: `no-qa-changes`
     */
    readonly labelToApplyWhenNoDiffPresent: string;

    /**
     * Duration in seconds for the assumed role to be valid
     * Defaut value: `900`
     */
    readonly roleDuration?: number;
  }

  export interface EnvToDiff extends BaseEnvToDiff {
    /**
     * String to search for stacks to diff
     * Example: `Qa`, 'Staging', 'Prod'
     */
    readonly stackSearchString: string;
  }

  export interface ExplicitStacksEnvToDiff extends BaseEnvToDiff {
    /**
     * Explicit stacks given instead of using stackSearchString to find stacks via cdk ls
     * Example: `['stack1', 'stack2']`
     */
    readonly stacks: string[];
  }

  export interface CDKDiffOptionsConfig extends OptionalNodeVersion {
    /**
     * Collection of environments to cdk diff
     */
    readonly envsToDiff: (EnvToDiff | ExplicitStacksEnvToDiff)[];

    /**
     * Detrmines if the OIDC role stack should be created
     */
    readonly createOidcRoleStack?: boolean;

    /**
     * Which package manager is being used?
     * @default - the packageManager in the project
     */
    readonly packageManager?: javascript.NodePackageManager;

    /**
     * What version of pnpm?
     * @default - 9 if you are using PNPM
     */
    readonly pnpmVersion?: string;
  }

  export function getCDKDiffOptions(options?: CDKDiffOptionsConfig) {
    return options;
  }

  export function addCdkDiffWorkflowYml(
    project: clickupCdk.ClickUpCdkTypeScriptApp,
    options: CDKDiffOptionsConfig,
    override?: any,
  ): void {
    new YamlFile(project, '.github/workflows/cdk-diff.yml', {
      obj: {
        ...createCdkDiffWorkflow({
          nodeVersion: project.workflowNodeVersion,
          packageManager: project.package.packageManager,
          pnpmVersion: project.package.pnpmVersion,
          ...options,
        }),
        ...override,
      },
    });
  }

  export function AddCdkLogParserDependency(pkg: NodePackage) {
    pkg.addDevDeps('@time-loop/cdk-log-parser');
  }

  export function addOidcRoleStack(project: clickupCdk.ClickUpCdkTypeScriptApp): void {
    new SampleFile(project, `${project.srcdir}/github-actions-oidc-permissions.ts`, {
      contents: `import { core } from '@time-loop/cdk-library';
import { aws_iam, Stage } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Namer } from 'multi-convention-namer';

export class GitHubActionsOIDCPermissions extends core.Stack {
  static asStage(scope: Construct, stageName: string, stageProps: core.StageProps): Stage {
    return new (class extends Stage {
      constructor() {
        super(scope, stageName, stageProps);
        new GitHubActionsOIDCPermissions(this, stageProps);
      }
    })();
  }
  constructor(scope: Construct, props: core.StackProps) {
    const projectName = '${project.name}';
    let id = new Namer([...projectName.split('-'), 'github', 'actions']);
    super(scope, id, props);

    const githubActionsRoleName = id.addSuffix(['role']).kebab;
    const githubActionsRole = new aws_iam.Role(this, githubActionsRoleName, {
      roleName: githubActionsRoleName,
      assumedBy: new aws_iam.FederatedPrincipal(
        \`arn:aws:iam::\${this.account}:oidc-provider/token.actions.githubusercontent.com\`,
        {
          StringEquals: {
            'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
          },
          StringLike: {
            'token.actions.githubusercontent.com:sub': \`repo:time-loop/\${projectName}:*\`,
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    });

    const githubActionsPolicyName = id.addSuffix(['policy']).kebab;
    const githubActionsPolicy = new aws_iam.Policy(this, githubActionsPolicyName, {
      policyName: githubActionsPolicyName,
      statements: [
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.ALLOW,
          actions: ['cloudformation:Describe*', 'cloudformation:List*', 'cloudformation:Get*'],
          resources: ['*'],
        }),
      ],
    });

    // Attach IAM policy to IAM role
    githubActionsPolicy.attachToRole(githubActionsRole);
  }
}
`,
    });
  }
}
