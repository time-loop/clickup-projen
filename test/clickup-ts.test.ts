import { Testing, javascript } from 'projen';
import { clickupTs } from '../src';

describe('ClickUpTypeScriptProject', () => {
  describe('defaults', () => {
    const p = new clickupTs.ClickUpTypeScriptProject({
      name: '@time-loop/test',
      defaultReleaseBranch: 'main',
    });
    const synth = Testing.synth(p);
    // console.log(Object.getOwnPropertyNames(synth));
    [
      'codecov.yml',
      'package.json',
      '.github/workflows/upgrade-main.yml',
      '.mergify.yml',
      '.github/workflows/release.yml',
    ].forEach((f) => {
      test(f, () => {
        expect(synth[f]).toMatchSnapshot();
      });
    });
    test('prettier is enabled', () => {
      expect(p.prettier).toBeTruthy();
    });
    test('jest is enabled', () => {
      expect(p.jest).toBeTruthy();
    });
    test('min node version is 18', () => {
      expect(synth['package.json'].engines.node).toBe('>= 18.17.1');
    });
    // TODO: soooo many more tests need to be written here.
  });
  describe('name', () => {
    let envCache = process.env;

    beforeEach(() => {
      process.env = envCache;
    });

    test('should add prefix when missing', () => {
      jest.spyOn(console, 'log');
      const p = new clickupTs.ClickUpTypeScriptProject({
        name: 'missing-prefix',
        defaultReleaseBranch: 'main',
      });
      expect(p.name).toBe('@time-loop/missing-prefix');
    });

    test('should respect GITHUB_OWNER', () => {
      jest.spyOn(console, 'log');
      process.env.GITHUB_OWNER = 'fizzle';
      const p = new clickupTs.ClickUpTypeScriptProject({
        name: 'missing-prefix',
        defaultReleaseBranch: 'main',
      });
      expect(p.name).toBe('@fizzle/missing-prefix');
    });
  });
  describe('docgen', () => {
    const projenDocgenTaskName = 'docgen';
    const clickupDocgenTaskName = 'typedocDocgen';
    const commonOpts: clickupTs.ClickUpTypeScriptProjectOptions = {
      name: 'random-prefix',
      defaultReleaseBranch: 'main',
    };

    test('succeeds with defaults', () => {
      new clickupTs.ClickUpTypeScriptProject(commonOpts);
    });

    test('includes markdown plugin by default', () => {
      const project = new clickupTs.ClickUpTypeScriptProject(commonOpts);
      expect(project.deps.getDependency('typedoc-plugin-markdown')).toBeDefined();
    });

    test('excludes markdown plugin when html attr is set', () => {
      const project = new clickupTs.ClickUpTypeScriptProject({ ...commonOpts, docgenOptions: { html: true } });
      const getDep = () => {
        project.deps.getDependency('typedoc-plugin-markdown');
      };
      expect(getDep).toThrowError();
    });

    test('includes typedocDocgen task by default', () => {
      const project = new clickupTs.ClickUpTypeScriptProject(commonOpts);
      expect(project.tasks.tryFind(clickupDocgenTaskName)).toBeDefined();
      expect(project.tasks.tryFind(projenDocgenTaskName)).toBeUndefined();
    });

    test('contains no docgen task when docgen is false', () => {
      const project = new clickupTs.ClickUpTypeScriptProject({ ...commonOpts, docgen: false });
      [clickupDocgenTaskName, projenDocgenTaskName].forEach((task) =>
        expect(project.tasks.tryFind(task)).toBeUndefined(),
      );
    });

    test('creates typedoc config JSON file', () => {
      const project = new clickupTs.ClickUpTypeScriptProject(commonOpts);
      const matchedFiles = project.files.filter((file) => file.path === 'typedoc.json');
      expect(matchedFiles).toHaveLength(1);
    });

    test('respects configFilePath attr when set', () => {
      const testOverride = 'test.json';
      const project = new clickupTs.ClickUpTypeScriptProject({
        ...commonOpts,
        docgenOptions: { configFilePath: testOverride },
      });
      const matchedFiles = project.files.filter((file) => file.path === testOverride);
      expect(matchedFiles).toHaveLength(1);
    });
  });
  describe('pnpm', () => {
    const p = new clickupTs.ClickUpTypeScriptProject({
      name: '@time-loop/test',
      defaultReleaseBranch: 'main',
      packageManager: javascript.NodePackageManager.PNPM,
      pnpmVersion: '9',
    });
    const synth = Testing.synth(p);
    it('packageManager', () => {
      expect(synth['package.json'].packageManager).toBe('pnpm@9.15.5');
    });

    const npmrcEntries = [
      'package-manager-strict=false',
      'manage-package-manager-versions=true',
      'use-node-version=22.13.1',
    ];
    it.each(npmrcEntries)('%s', (npmrcEntry) => {
      expect(synth['.npmrc']).toContain(npmrcEntry);
    });
  });
});

describe('normalizeName', () => {
  const envCache = process.env;

  beforeEach(() => {
    process.env = envCache;
    delete process.env.GITHUB_OWNER;
  });

  test('should silently pass when prefix is present', () => {
    const logSpy = jest.spyOn(console, 'log');
    const name = clickupTs.normalizeName('@time-loop/has-prefix');
    expect(name).toBe('@time-loop/has-prefix');
    expect(logSpy).not.toHaveBeenCalled();
  });

  test('should add prefix and complain when missing', () => {
    const logSpy = jest.spyOn(console, 'log');
    const name = clickupTs.normalizeName('missing-prefix');
    expect(name).toBe('@time-loop/missing-prefix');
    expect(logSpy).toHaveBeenCalledWith(
      'Adding mandatory prefix @time-loop/ to name. New name: @time-loop/missing-prefix',
    );
  });

  test('should respect GITHUB_OWNER', () => {
    const logSpy = jest.spyOn(console, 'log'); // silence console.log
    process.env.GITHUB_OWNER = 'fizzle';
    const name = clickupTs.normalizeName('missing-prefix');
    expect(name).toBe('@fizzle/missing-prefix');
    expect(logSpy).toHaveBeenCalled();
  });
});
