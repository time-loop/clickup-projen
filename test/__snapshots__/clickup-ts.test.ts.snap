// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickUpTypeScriptProject defaults .github/workflows/release.yml 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: release
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
concurrency:
  group: \${{ github.workflow }}
  cancel-in-progress: false
jobs:
  release:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    outputs:
      latest_commit: \${{ steps.git_remote.outputs.latest_commit }}
      tag_exists: \${{ steps.check_tag_exists.outputs.exists }}
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          fetch-depth: 0
      - name: Set git identity
        run: |-
          git config user.name "github-actions"
          git config user.email "github-actions@github.com"
      - name: GitHub Packages authorization
        env:
          NPM_TOKEN: \${{ secrets.ALL_PACKAGE_READ_TOKEN }}
        run: |-
          cat > ~/.npmrc <<EOF
          //npm.pkg.github.com/:_authToken=\${NPM_TOKEN}
          @time-loop:registry=https://npm.pkg.github.com/
          EOF
      - name: Make cdk-ecr-deployment sane
        run: export FORCE_PREBUILT_LAMBDA=1
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 22.13.1
      - name: Install dependencies
        run: yarn install --check-files --frozen-lockfile
      - name: release
        run: npx projen release
      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@v4
        with:
          token: \${{ secrets.CODECOV_TOKEN }}
          directory: coverage
      - name: Check if version has already been tagged
        id: check_tag_exists
        run: |-
          TAG=$(cat dist/releasetag.txt)
          ([ ! -z "$TAG" ] && git ls-remote -q --exit-code --tags origin $TAG && (echo "exists=true" >> $GITHUB_OUTPUT)) || (echo "exists=false" >> $GITHUB_OUTPUT)
          cat $GITHUB_OUTPUT
      - name: Check for new commits
        id: git_remote
        run: |-
          echo "latest_commit=$(git ls-remote origin -h \${{ github.ref }} | cut -f1)" >> $GITHUB_OUTPUT
          cat $GITHUB_OUTPUT
      - name: Backup artifact permissions
        if: \${{ steps.git_remote.outputs.latest_commit == github.sha }}
        run: cd dist && getfacl -R . > permissions-backup.acl
        continue-on-error: true
      - name: Upload artifact
        if: \${{ steps.git_remote.outputs.latest_commit == github.sha }}
        uses: actions/upload-artifact@v4.4.0
        with:
          name: build-artifact
          path: dist
          overwrite: true
  release_github:
    name: Publish to GitHub Releases
    needs:
      - release
      - release_npm
    runs-on: ubuntu-latest
    permissions:
      contents: write
    if: needs.release.outputs.tag_exists != 'true' && needs.release.outputs.latest_commit == github.sha
    steps:
      - uses: actions/setup-node@v4
        with:
          node-version: 22.13.1
      - name: Download build artifacts
        uses: actions/download-artifact@v4
        with:
          name: build-artifact
          path: dist
      - name: Restore build artifact permissions
        run: cd dist && setfacl --restore=permissions-backup.acl
        continue-on-error: true
      - name: Release
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
          GITHUB_REPOSITORY: \${{ github.repository }}
          GITHUB_REF: \${{ github.sha }}
        run: errout=$(mktemp); gh release create $(cat dist/releasetag.txt) -R $GITHUB_REPOSITORY -F dist/changelog.md -t $(cat dist/releasetag.txt) --target $GITHUB_REF 2> $errout && true; exitcode=$?; if [ $exitcode -ne 0 ] && ! grep -q "Release.tag_name already exists" $errout; then cat $errout; exit $exitcode; fi
  release_npm:
    name: Publish to npm
    needs: release
    runs-on: ubuntu-latest
    permissions:
      contents: read
      packages: write
    if: needs.release.outputs.tag_exists != 'true' && needs.release.outputs.latest_commit == github.sha
    steps:
      - uses: actions/setup-node@v4
        with:
          node-version: 22.13.1
      - name: Download build artifacts
        uses: actions/download-artifact@v4
        with:
          name: build-artifact
          path: dist
      - name: Restore build artifact permissions
        run: cd dist && setfacl --restore=permissions-backup.acl
        continue-on-error: true
      - name: Release
        env:
          NPM_DIST_TAG: latest
          NPM_REGISTRY: npm.pkg.github.com
          NPM_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        run: npx -p publib@latest publib-npm
  send_release_event_to_slack:
    name: Send Release Alert to Slack
    needs: release
    runs-on: ubuntu-latest
    permissions:
      contents: read
    if: needs.release.outputs.latest_commit == github.sha
    steps:
      - name: Download build artifacts
        uses: actions/download-artifact@v4
        with:
          name: build-artifact
          path: dist
      - name: Get version
        id: event_metadata
        run: echo "release_tag=$(cat dist/releasetag.txt)" >> $GITHUB_OUTPUT
      - name: Send Slack webhook event
        uses: rtCamp/action-slack-notify@12e36fc18b0689399306c2e0b3e0f2978b7f1ee7
        env:
          SLACK_TITLE: \${{ github.repository }}@\${{ steps.event_metadata.outputs.release_tag }} released!
          SLACK_MESSAGE: "View the release notes here: https://github.com/\${{ github.repository }}/releases/tag/\${{ steps.event_metadata.outputs.release_tag }}"
          SLACK_WEBHOOK: \${{ secrets.PROJEN_RELEASE_SLACK_WEBHOOK }}
          SLACK_FOOTER: ""
          SLACK_COLOR: success
          MSG_MINIMAL: "true"
"
`;

exports[`ClickUpTypeScriptProject defaults .github/workflows/upgrade-main.yml 1`] = `undefined`;

exports[`ClickUpTypeScriptProject defaults .mergify.yml 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

queue_rules:
  - name: default
    update_method: merge
    conditions:
      - "#approved-reviews-by>=1"
      - -label~=(do-not-merge)
      - status-success=build
    merge_method: squash
    commit_message_template: |-
      {{ title }} (#{{ number }})

      {{ body }}
pull_request_rules:
  - name: Automatic merge on approval and successful build
    actions:
      delete_head_branch: {}
      queue:
        name: default
    conditions:
      - "#approved-reviews-by>=1"
      - -label~=(do-not-merge)
      - status-success=build
"
`;

exports[`ClickUpTypeScriptProject defaults codecov.yml 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

coverage:
  precision: 2
  round: down
  status:
    project:
      default:
        target: auto
        threshold: 10%
        paths:
          - src
        if_ci_failed: error
        only_pulls: true
    patch:
      default:
        base: auto
        target: auto
        threshold: 10%
        paths:
          - src
        if_ci_failed: error
        only_pulls: true
parsers:
  gcov:
    branch_detection:
      conditional: yes
      loop: yes
      method: no
      macro: no
comment:
  layout: reach,diff,flags,files,footer
  behavior: default
  require_changes: no
"
`;

exports[`ClickUpTypeScriptProject defaults package.json 1`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
  "author": {
    "name": "ClickUp",
    "organization": true,
  },
  "dependencies": {
    "@time-loop/clickup-projen": "*",
  },
  "devDependencies": {
    "@types/jest": "*",
    "@types/node": "^18",
    "@typescript-eslint/eslint-plugin": "^8",
    "@typescript-eslint/parser": "^8",
    "commit-and-tag-version": "^12",
    "constructs": "^10.0.0",
    "esbuild": "*",
    "eslint": "^9",
    "eslint-config-prettier": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "eslint-plugin-prettier": "*",
    "jest": "*",
    "jest-junit": "^16",
    "jsii-release": "*",
    "prettier": "*",
    "projen": "*",
    "ts-jest": "*",
    "ts-node": "^10",
    "typedoc": "*",
    "typedoc-plugin-markdown": "*",
    "typescript": "*",
  },
  "engines": {
    "node": ">= 18.17.1",
  },
  "jest": {
    "clearMocks": true,
    "collectCoverage": true,
    "collectCoverageFrom": [
      "src/**/*.ts",
    ],
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": [
      "/node_modules/",
    ],
    "coverageProvider": "v8",
    "coverageReporters": [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "reporters": [
      "default",
      [
        "jest-junit",
        {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": [
      "<rootDir>/@(src|test)/**/*(*.)@(spec|test).ts?(x)",
      "<rootDir>/@(src|test)/**/__tests__/**/*.ts?(x)",
    ],
    "testPathIgnorePatterns": [
      "/node_modules/",
    ],
    "transform": {
      "^.+\\.[t]sx?$": [
        "ts-jest",
        {
          "tsconfig": "tsconfig.dev.json",
        },
      ],
    },
    "watchPathIgnorePatterns": [
      "/node_modules/",
    ],
  },
  "license": "UNLICENSED",
  "main": "lib/index.js",
  "name": "@time-loop/test",
  "publishConfig": {
    "registry": "https://npm.pkg.github.com/",
  },
  "scripts": {
    "build": "npx projen build",
    "bump": "npx projen bump",
    "clobber": "npx projen clobber",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "eject": "npx projen eject",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "release": "npx projen release",
    "test": "npx projen test",
    "test:watch": "npx projen test:watch",
    "typedocDocgen": "npx projen typedocDocgen",
    "unbump": "npx projen unbump",
    "update-projen": "npx projen update-projen",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "types": "lib/index.d.ts",
  "version": "0.0.0",
}
`;
